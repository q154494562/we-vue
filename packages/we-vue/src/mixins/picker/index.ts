// Mixins
import Toggleable, { factory as ToggleableFactory } from '../toggleable'

// Utils
import mixins, { ExtractVue } from '@/utils/mixins'

export default mixins<
  ExtractVue<[typeof Toggleable]>
>(
  ToggleableFactory('visible', 'update:visible')
).extend({
  name: 'picker',

  props: {
    confirmText: {
      type: String,
      default: '确定',
    },
    cancelText: {
      type: String,
      default: '取消',
    },
    closeOnClickMask: {
      type: Boolean,
      default: true,
    },
  },

  methods: {
    open (): void {
      this.isActive = true
    },

    close (): void {
      this.isActive = false
    },
  },
})
