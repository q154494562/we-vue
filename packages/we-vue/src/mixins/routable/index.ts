import Vue from 'vue'

export default Vue.extend({
  name: 'routable',

  props: {
    href: String,
    replace: Boolean,
    to: [String, Object],
  },

  methods: {
    routeLink () {
      const { to, href, $router, replace } = this as any
      if (to && $router) {
        $router[replace ? 'replace' : 'push'](to)
      } else if (href) {
        replace ? location.replace(href) : location.href = href
      }
    },
  },
})
