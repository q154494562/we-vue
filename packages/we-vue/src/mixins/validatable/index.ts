import Vue from 'vue'

export default Vue.extend({
  name: 'validatable',

  props: {
    value: {
      required: false,
    },
  },

  data () {
    return {
      lazyValue: this.value,
      valid: false,
    }
  },
})
