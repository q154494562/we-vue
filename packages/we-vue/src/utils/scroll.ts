type ElementOrWindow = HTMLElement | Window

/**
 * 找到最近的触发滚动事件的元素
 * @param {Element} element
 * @param {Element} rootParent
 * @returns {Element | window}
 */
export function getScrollEventTarget (element: HTMLElement, rootParent: ElementOrWindow = window) {
  let node = element
  while (
    node &&
    node.tagName !== 'HTML' &&
    node.tagName !== 'BODY' &&
    node.nodeType === 1 &&
    node !== rootParent
  ) {
    const { overflowY } = window.getComputedStyle(node)
    if (overflowY === 'scroll' || overflowY === 'auto') {
      return node
    }
    node = <HTMLElement>node.parentNode
  }
  return rootParent
}

// 获取滚动高度
export function getScrollTop (element: ElementOrWindow): number {
  return 'scrollTop' in element ? element.scrollTop : element.pageYOffset
}

// 设置滚动高度
export function setScrollTop (element: ElementOrWindow, value: number) {
  'scrollTop' in element ? (element.scrollTop = value) : element.scrollTo(element.scrollX, value)
}

// 获取元素距离顶部高度
export function getElementTop (element: ElementOrWindow) {
  return (
    (element === window ? 0 : (element as HTMLElement).getBoundingClientRect().top) + getScrollTop(window)
  )
}

export function getVisibleHeight (element: ElementOrWindow) {
  return element === window
    ? element.innerHeight
    : (element as HTMLElement).getBoundingClientRect().height
}
