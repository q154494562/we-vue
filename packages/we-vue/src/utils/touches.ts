function getTouch (event: TouchEvent): Touch {
  return event.changedTouches[0] || event.touches[0]
}

export { getTouch }
