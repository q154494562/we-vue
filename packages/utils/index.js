import Vue from 'vue'
import { getTouch } from './touches'

const isServer = Vue.prototype.$isServer

function isDef (value) {
  return value !== undefined && value !== null
}

function isObj (x) {
  const type = typeof x
  return x !== null && (type === 'object' || type === 'function')
}

export {
  isServer,
  getTouch,
  isDef,
  isObj
}
