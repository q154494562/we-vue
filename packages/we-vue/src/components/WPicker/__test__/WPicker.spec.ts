import { mount, Wrapper } from '@vue/test-utils'
import Picker from '../'
import PickerColumn from '../WPickerColumn'
import { slowVerticalDrag } from '@/test/unit/utils'
import { ExtractVue } from '@/utils/mixins'

const testSingleColumn = [
  {
    options: [1, 2, 3],
  },
]

const testMultiColumn = [
  {
    options: [1, 2, 3],
  },
  {
    options: ['yes', 'no'],
  },
]

type PickerWrapper = Wrapper<ExtractVue<typeof Picker>>
type PickerColumnWrapper = Wrapper<ExtractVue<typeof PickerColumn>>

describe('picker', () => {
  test('create', () => {
    const wrapper = mount(Picker, {
      propsData: {
        visible: true,
      },
    })

    expect(wrapper.name()).toBe('w-picker')

    expect(wrapper.html()).toMatchSnapshot()
  })

  test('create a single-column picker', () => {
    const wrapper = mount(Picker, {
      attachToDocument: true,
      propsData: {
        visible: true,
        columns: testSingleColumn,
      },
    })

    expect(wrapper.findAll(PickerColumn)).toHaveLength(1)

    expect(wrapper.vm.getColumnOptions(0)).toHaveLength(3)
    expect(wrapper.vm.getValues()).toEqual([1])

    expect(wrapper.html()).toMatchSnapshot()
  })

  test('create a multi-column picker', () => {
    const wrapper = mount(Picker, {
      attachToDocument: true,
      propsData: {
        visible: true,
        columns: testMultiColumn,
      },
    })

    expect(wrapper.findAll(PickerColumn)).toHaveLength(2)

    expect(wrapper.vm.getColumnOptions(0)).toHaveLength(3)
    expect(wrapper.vm.getColumnOptions(1)).toHaveLength(2)
    expect(wrapper.vm.getValues()).toEqual([1, 'yes'])
    expect(wrapper.html()).toMatchSnapshot()
  })

  test('getColumnValue method', () => {
    const wrapper = mount(Picker, {
      attachToDocument: true,
      propsData: {
        visible: true,
        columns: testSingleColumn,
      },
    })

    expect(wrapper.vm.getColumnValue(0)).toEqual(1)
  })

  test('getColumnOptions method', () => {
    const wrapper = mount(Picker, {
      attachToDocument: true,
      propsData: {
        visible: true,
        columns: testMultiColumn,
      },
    })

    expect(wrapper.vm.getColumnOptions(0)).toEqual([1, 2, 3])
    expect(wrapper.vm.getColumnOptions(1)).toEqual(['yes', 'no'])
  })

  test('test setColumnValue method', () => {
    const wrapper = mount(Picker, {
      attachToDocument: true,
      propsData: {
        visible: true,
        columns: testMultiColumn,
      },
    })

    wrapper.vm.setColumnOptions(0, [1, 2, 3])

    expect(wrapper.vm.getColumnOptions(0)).toEqual([1, 2, 3])
  })

  test('test getValues method', () => {
    const wrapper = mount(Picker, {
      attachToDocument: true,
      propsData: {
        visible: true,
        columns: testSingleColumn,
      },
    })

    expect(wrapper.vm.getValues()).toEqual([1])
  })

  test('test setValues method', () => {
    const wrapper = mount(Picker, {
      attachToDocument: true,
      propsData: {
        visible: true,
        columns: [
          {
            options: [1, 2, 3],
            default: 0,
          },
        ],
      },
    })

    wrapper.vm.setValues([2])

    expect(wrapper.vm.getValues()).toEqual([2])
  })

  test('test getColumnIndex method', () => {
    const wrapper = mount(Picker, {
      attachToDocument: true,
      propsData: {
        visible: true,
        columns: testMultiColumn,
      },
    })

    expect(wrapper.vm.getColumnIndex(0)).toBe(0)
    expect(wrapper.vm.getColumnIndex(1)).toBe(0)
  })

  test('test setColumnIndex method', () => {
    const wrapper = mount(Picker, {
      attachToDocument: true,
      propsData: {
        visible: true,
        columns: testMultiColumn,
      },
    })

    wrapper.vm.setColumnIndex(0, 1)
    wrapper.vm.setColumnIndex(1, 1)

    expect(wrapper.vm.getIndexes()).toEqual([1, 1])
  })

  test('test getIndexes method', () => {
    const wrapper = mount(Picker, {
      attachToDocument: true,
      propsData: {
        visible: true,
        columns: testMultiColumn,
      },
    })

    expect(wrapper.vm.getIndexes()).toEqual([0, 0])
  })

  test('test setIndexes method', () => {
    const wrapper = mount(Picker, {
      attachToDocument: true,
      propsData: {
        visible: true,
        columns: testMultiColumn,
      },
    })

    wrapper.vm.setIndexes([1, 1])

    expect(wrapper.vm.getIndexes()).toEqual([1, 1])
  })

  test('click cancel button', () => {
    const wrapper = mount(Picker, {
      attachToDocument: true,
      propsData: {
        'visible.sync': true,
        columns: [],
      },
    })

    wrapper
      .findAll('.weui-picker__action')
      .at(0)
      .trigger('click')
    expect(wrapper.vm.isActive).toBe(false)
    expect(wrapper.emitted().cancel).toBeTruthy()
  })

  test('click confirm button', () => {
    const wrapper = mount(Picker, {
      attachToDocument: true,
      propsData: {
        'visible.sync': true,
        columns: [],
      },
    })

    wrapper
      .findAll('.weui-picker__action')
      .at(1)
      .trigger('click')
    expect(wrapper.vm.isActive).toBe(false)
    expect(wrapper.emitted().confirm).toBeTruthy()
  })

  test('when column value changed, change event should be emitted', () => {
    const wrapper = mount(Picker, {
      attachToDocument: true,
      propsData: {
        visible: true,
        columns: [
          {
            options: [1, 2, 3],
            default: 0,
          },
        ],
      },
    })

    wrapper
      .findAll(PickerColumn)
      .at(0)
      .vm.$emit('change', 0)

    expect(wrapper.emitted().change).toBeTruthy()
  })

  test('value watcher', () => {
    const wrapper = mount(Picker, {
      attachToDocument: true,
      propsData: {
        visible: true,
        columns: [
          {
            options: [1, 2, 3],
            default: 0,
          },
        ],
      },
    })

    const spy = jest.spyOn(wrapper.vm, 'setValues')

    wrapper.setProps({
      value: [2],
    })

    expect(spy).toHaveBeenCalled()
  })
})

describe('picker-column', () => {
  // test('create', () => {
  //   const wrapper = mount(PickerColumn, {
  //     parentComponent: Picker
  //   })
  //
  //   expect(wrapper.name()).toBe('w-picker-column')
  // })

  // test('create using object-array values', () => {
  //   const options = [
  //     {
  //       label: 'label1',
  //       value: 'value1'
  //     },
  //     {
  //       label: 'label2',
  //       value: 'value2'
  //     },
  //     {
  //       label: 'label3',
  //       value: 'value3'
  //     }
  //   ]
  //
  //   const wrapper = mount(PickerColumn, {
  //     parentComponent: Picker,
  //     propsData: {
  //       options: options,
  //       valueKey: 'value'
  //     }
  //   })
  //
  //   expect(wrapper.findAll('.weui-picker__item').length).toBe(3)
  // })

  // test('render a divider slot', () => {
  //   const wrapper = mount(PickerColumn, {
  //     parentComponent: Picker,
  //     propsData: {
  //       divider: true,
  //       content: '-'
  //     }
  //   })
  //
  //   expect(wrapper.text()).toBe('-')
  // })

  test(
    'drag slot',
    () => {
      const wrapper = mount(Picker, {
        attachToDocument: true,
        propsData: {
          visible: true,
          columns: testSingleColumn,
        },
      })

      const columnWrapper = wrapper.find(PickerColumn) as PickerColumnWrapper

      slowVerticalDrag(columnWrapper, 0, -34)
      expect(columnWrapper.vm.currentIndex).toBe(1)

      slowVerticalDrag(columnWrapper, 0, -34)
      expect(columnWrapper.vm.currentIndex).toBe(2)

      // this will out of range
      slowVerticalDrag(columnWrapper, 0, -34)
      expect(columnWrapper.vm.currentIndex).toBe(2)

      slowVerticalDrag(columnWrapper, 0, 34)
      expect(columnWrapper.vm.currentIndex).toBe(1)

      // this will out of range
      slowVerticalDrag(columnWrapper, 0, 100)
      expect(columnWrapper.vm.currentIndex).toBe(0)
    },
    15000
  )

  test('click slot to change the current-value', () => {
    const wrapper = mount(Picker, {
      attachToDocument: true,
      propsData: {
        visible: true,
        columns: [
          {
            options: [1, 2, 3],
            defaultIndex: 0,
          },
        ],
      },
    })

    const columnWrapper = wrapper.find(PickerColumn) as PickerColumnWrapper

    const option1 = wrapper.findAll('.weui-picker__item').at(1)
    option1.trigger('click')

    expect(columnWrapper.vm.currentIndex).toBe(1)
  })

  test('divider pickerSlot', () => {
    const wrapper = mount(Picker, {
      attachToDocument: true,
      propsData: {
        visible: true,
        columns: [
          {
            divider: true,
          },
        ],
      },
    })

    const pickerColumnWrapper = wrapper.find(PickerColumn)

    pickerColumnWrapper.setData({
      currentIndex: 1,
    })

    // divider should not emit change event when currentIndex changed
    expect(pickerColumnWrapper.emitted().change).toBeFalsy()
  })

  test('index should be adjust to a suitable value when it is exceeded ot disabled', () => {
    const wrapper = mount(Picker, {
      attachToDocument: true,
      propsData: {
        visible: true,
        columns: [
          {
            options: [
              {
                text: 1,
              },
              {
                text: 2,
              },
              {
                text: 3,
                disabled: true,
              },
            ],
            defaultIndex: 0,
          },
        ],
      },
    })

    const pickerColumnWrapper = wrapper.find(PickerColumn) as PickerColumnWrapper

    pickerColumnWrapper.vm.setIndex(4)

    // divider should not emit change event when currentIndex changed
    expect(pickerColumnWrapper.vm.currentIndex).toBe(1)
  })

  test('watch defaultIndex', () => {
    const wrapper = mount(Picker, {
      attachToDocument: true,
      propsData: {
        visible: true,
        columns: [
          {
            options: [1, 2, 3],
            defaultIndex: 0,
          },
        ],
      },
    })

    const pickerColumnWrapper = wrapper.find(PickerColumn) as PickerColumnWrapper

    pickerColumnWrapper.setProps({
      defaultIndex: 1,
    })

    expect(pickerColumnWrapper.vm.currentIndex).toBe(1)
  })

  test('test setValue method', () => {
    const wrapper: PickerWrapper = mount(Picker, {
      attachToDocument: true,
      propsData: {
        visible: true,
        columns: [
          {
            options: [1, 2, 3],
            defaultIndex: 0,
          },
        ],
      },
    })

    const pickerColumnWrapper = wrapper.find(PickerColumn) as PickerColumnWrapper

    const spy = jest.spyOn(pickerColumnWrapper.vm, 'setIndex')

    pickerColumnWrapper.vm.setValue(3)

    expect(spy).toHaveBeenCalledWith(2)
  })
})
