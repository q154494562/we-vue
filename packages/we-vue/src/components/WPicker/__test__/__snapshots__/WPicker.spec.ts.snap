// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`picker create 1`] = `
<div>
  <div class="weui-mask"></div>
  <div class="weui-picker">
    <div class="weui-picker__hd">
      <div class="weui-picker__action">取消</div>
      <div class="weui-picker__action">确定</div>
    </div>
    <div class="weui-picker__bd" style="height: 238px;"></div>
  </div>
</div>
`;

exports[`picker create a multi-column picker 1`] = `
<div>
  <div class="weui-mask"></div>
  <div class="weui-picker">
    <div class="weui-picker__hd">
      <div class="weui-picker__action">取消</div>
      <div class="weui-picker__action">确定</div>
    </div>
    <div class="weui-picker__bd" style="height: 238px;">
      <div class="weui-picker__group">
        <div class="weui-picker__mask" style="background-size: 100% 102px;"></div>
        <div class="weui-picker__indicator" style="top: 102px;"></div>
        <div class="weui-picker__content" style="transform: translate3d(0, 102px, 0);">
          <div class="weui-picker__item">1</div>
          <div class="weui-picker__item">2</div>
          <div class="weui-picker__item">3</div>
        </div>
      </div>
      <div class="weui-picker__group">
        <div class="weui-picker__mask" style="background-size: 100% 102px;"></div>
        <div class="weui-picker__indicator" style="top: 102px;"></div>
        <div class="weui-picker__content" style="transform: translate3d(0, 102px, 0);">
          <div class="weui-picker__item">yes</div>
          <div class="weui-picker__item">no</div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`picker create a single-column picker 1`] = `
<div>
  <div class="weui-mask"></div>
  <div class="weui-picker">
    <div class="weui-picker__hd">
      <div class="weui-picker__action">取消</div>
      <div class="weui-picker__action">确定</div>
    </div>
    <div class="weui-picker__bd" style="height: 238px;">
      <div class="weui-picker__group">
        <div class="weui-picker__mask" style="background-size: 100% 102px;"></div>
        <div class="weui-picker__indicator" style="top: 102px;"></div>
        <div class="weui-picker__content" style="transform: translate3d(0, 102px, 0);">
          <div class="weui-picker__item">1</div>
          <div class="weui-picker__item">2</div>
          <div class="weui-picker__item">3</div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
